src/lambda/commands/uploadLambda.ts

```diff
@@ -179,26 +189,49 @@ class UploadLambdaWizard extends Wizard<UploadLambdaWizardState> {
             }
         })
 
-        this.form.directoryUploadType.bindPrompter(() => createDirectoryUploadPrompter(), {
-            showWhen: ({ uploadType }) => uploadType === 'directory',
+        this.form.lambda.name.bindPrompter(state => {
+            // invoking from the command palette passes no arguments
+            if (!invokePath) {
+                if (state.uploadType === 'directory') {
+                    return createFunctionNamePrompter(state.lambda!.region!, state.targetUri)
+                } else {
+                    // if a zip file is chosen, pass the parent directory
+                    return createFunctionNamePrompter(
+                        state.lambda!.region!,
+                        vscode.Uri.parse(path.dirname(state.targetUri!.fsPath))
+                    )
+                }
+            } else {
+                return createFunctionNamePrompter(state.lambda!.region!, invokePath)
+            }
         })
+        if (lambda) {
+            this.form.directoryBuildType.bindPrompter(() => createBuildPrompter(), {
+                showWhen: ({ uploadType }) => uploadType === 'directory',
+            })
+        } else {
+            this.form.directoryBuildType.setDefault('zip')
+        }
 
-        this.form.confirmedDeploy.bindPrompter(() => createConfirmDeploymentPrompter(lambda))
+        this.form.confirmedDeploy.bindPrompter(state => createConfirmDeploymentPrompter(state.lambda!))
```

Comment:
there are a lot of `!` assumptions here. usually that can be avoided by checking once at the top, and returning or throwing an error as appropriate.

Reply:
Need to update the wizard code to support declaration of dependencies. The assertions are fine so long as the code is tested though obviously it'd be better to not have them at all.